api/v1beta2/cluster_types.go

```diff
@@ -586,6 +575,24 @@ type Topology struct {
 	Variables []ClusterVariable `json:"variables,omitempty"`
 }
 
+// ClusterClassRef is the ref to the ClusterClass that should be used for the topology.
+type ClusterClassRef struct {
+	// name is the name of the ClusterClass that should be used for the topology.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Name string `json:"name"`
```

Comment:
This is a CRD so would be a DNS 1123 subdomain validation, eventually (1.33 minimum) we can use CEL for this, but for now, can we at least update the godoc to explain the validation and add a pattern validation?
```
// Name must be at most 253 characters in length and must consist only of alphanumeric characters,
// hyphens (-) and periods (.), and must start and end with an alphanumeric character.
```

```suggestion
	// +kubebuilder:validation:Pattern=`^[a-z0-9](?:[-a-z0-9]*[a-z0-9])?(?:\.[a-z0-9](?:[-a-z0-9]*[a-z0-9])?)*$`
	Name string `json:"name"`
```

Reply:
Are you sure this is the right regex? I think up until now we only used that one for a Namespace

Reply:
And the namespace validation in Kubernetes seems to use the DNS 1123 label regexp.

So either this one is wrong or the one we used for Namespace fields is wrong.

Kubernetes error messages:

ClusterClass name:
```
The ClusterClass "quick-startasdf252bgfn.?" is invalid: metadata.name: Invalid value: "quick-startasdf252bgfn.?": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

Namespace name:
```
* metadata.name: Invalid value: "asdfasfadsfadsf.sdf?": a lowercase RFC 1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')
```

Reply:
Okay just saw your next comment.. But anyway. You're sure about this one, right? :D

Reply:
> except added non-capturing groups, which has no effect on functionality.

@JoelSpeed Any reason why we shouldn't just use the upstream regexp if there is no difference in functionality?

Reply:
Yeah. I really don't know why. I only have this info https://github.com/kubernetes-sigs/cluster-api/pull/11352#discussion_r1913130816

(cc @Danil-Grigorev)

Reply:
@JoelSpeed Looks like I don't need the additional backslash here `\\.`. 

Hope the regexp are now correct, pushed another commit